File: OsmAnd-java/src/net/osmand/data/Street.java
package net.osmand.data;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import net.osmand.util.Algorithms;


public class Street extends MapObject {
	
	protected List<Building> buildings = new ArrayList<Building>(); 
	protected List<Street> intersectedStreets = null;
	protected final City city;

	public Street(City city) {
		this.city = city;
	}
	
	public void addBuilding(Building building){
		buildings.add(building);
	}
	
	public List<Street> getIntersectedStreets() {
		if(intersectedStreets == null){
			return Collections.emptyList();
		}
		return intersectedStreets;
	}
	
	public void addIntersectedStreet(Street s){
		if(intersectedStreets == null) {
			intersectedStreets = new ArrayList<Street>();
		}
		intersectedStreets.add(s);
	}
	
	public void addBuildingCheckById(Building building){
		for(Building b : buildings) {
			if(b.getId().longValue() == building.getId().longValue()){
				return;
			}
		}
		buildings.add(building);
	}
	
	public List<Building> getBuildings() {
		return buildings;
	}
	
	
	
	public City getCity() {
		return city;
	}
	
	public void sortBuildings(){
		Collections.sort(buildings, new Comparator<Building>(){
			@Override
			public int compare(Building o1, Building o2) {
				String s1 = o1.getName();
				String s2 = o2.getName();
				int i1 = Algorithms.extractFirstIntegerNumber(s1);
				int i2 = Algorithms.extractFirstIntegerNumber(s2);
				if(i1 == i2) {
					String t1 = Algorithms.extractIntegerSuffix(s1);
					String t2 = Algorithms.extractIntegerSuffix(s2);
					return t1.compareTo(t2);
				}
				return i1 - i2;
			}
		});
	}
	
	/// GENERATION
	
	
	public void mergeWith(Street street) {
		buildings.addAll(street.getBuildings());
		copyNames(street);
	}

	public String getNameWithoutCityPart(String lang) {
		String nm = getName(lang);
		int t = nm.lastIndexOf('(');
		if (t > 0) {
			return nm.substring(0, t);
		}
		return nm;
	}


}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/data/Street.kt
package net.osmand.shared.data

import net.osmand.shared.util.KAlgorithms
import net.osmand.shared.util.KMapUtils

/**
 * A street of a [City] in the address section, with the buildings and crossing streets read for it.
 *
 * A copy of `net.osmand.data.Street` in OsmAnd-java, which stays there for android and tools. Left
 * out: the JSON export, which only writes test data from OsmAnd-java.
 */
class Street(private val city: City?) : MapObject() {

	private val buildings: MutableList<Building> = ArrayList()
	private var buildingsByIdCache: MutableMap<String, Building>? = null
	private var intersectedStreets: MutableList<Street>? = null

	fun addBuilding(building: Building) {
		buildings.add(building)
	}

	fun getIntersectedStreets(): List<Street> = intersectedStreets ?: emptyList()

	fun addIntersectedStreet(s: Street) {
		var streets = intersectedStreets
		if (streets == null) {
			streets = ArrayList()
			intersectedStreets = streets
		}
		streets.add(s)
	}

	override fun getBbox31(): IntArray? {
		val bb = getBboxPoints() ?: return null
		return intArrayOf(
			KMapUtils.get31TileNumberX(bb.left), KMapUtils.get31TileNumberY(bb.top),
			KMapUtils.get31TileNumberX(bb.right), KMapUtils.get31TileNumberY(bb.bottom)
		)
	}

	fun getBboxPoints(): KQuadRect? {
		val ll = getLocation() ?: return null
		val qr = getMinBbox(ll)
		if (buildings.isEmpty()) {
			// use intersected streets however it's much larger
			for (s in getIntersectedStreets()) {
				val l2 = s.getLocation()
				if (l2 != null) {
					qr.include(l2.longitude, l2.latitude)
				}
			}
		}
		for (b in buildings) {
			val l2 = b.getLocation()
			if (l2 != null) {
				qr.include(l2.longitude, l2.latitude)
			}
		}
		return qr
	}

	fun addBuildingCheckById(building: Building) {
		var cache = buildingsByIdCache
		if (cache == null) {
			cache = HashMap()
			for (b in buildings) {
				cache[b.getId().toString() + " " + b.getFullName()] = b
			}
			buildingsByIdCache = cache
		}
		val key = building.getId().toString() + " " + building.getFullName()
		if (cache.containsKey(key)) {
			return
		}
		cache[key] = building
		buildings.add(building)
	}

	fun getBuildings(): MutableList<Building> = buildings

	fun getCity(): City? = city

	fun sortBuildings() {
		buildings.sortWith { o1, o2 ->
			val s1 = o1.getName()
			val s2 = o2.getName()
			val i1 = KAlgorithms.extractFirstIntegerNumber(s1)
			val i2 = KAlgorithms.extractFirstIntegerNumber(s2)
			if (i1 == i2) {
				val t1 = KAlgorithms.extractIntegerSuffix(s1)
				val t2 = KAlgorithms.extractIntegerSuffix(s2)
				t1.compareTo(t2)
			} else {
				i1 - i2
			}
		}
	}

	/// GENERATION

	fun mergeWith(street: Street) {
		for (b in street.getBuildings()) {
			addBuildingCheckById(b)
		}
		copyNames(street)
	}

	fun getNameWithoutCityPart(lang: String?, transliterate: Boolean): String {
		val nm = getName(lang, transliterate)
		val t = nm.lastIndexOf('(')
		if (t > 0) {
			return nm.substring(0, t)
		}
		return nm
	}

	fun getNameCityPart(lang: String?, transliterate: Boolean): String {
		val nm = getName(lang, transliterate)
		var t = nm.lastIndexOf('(')
		if (t > 0) {
			var cityPart = nm.substring(t + 1)
			t = cityPart.lastIndexOf(')')
			if (t != -1) {
				cityPart = cityPart.substring(0, t)
			}
			return cityPart
		}
		return ""
	}
}
